### ISSUE TITLE
device integration

### ISSUE THREAD

**Issue by kkasravi** [kubeflow/kubeflow#2882]
**Table of Contents**  *generated with [DocToc](https://github.com/thlorenz/doctoc)*

- [Device Integration into Kubeflow](#device-integration-into-kubeflow)
  - [Scope](#scope)
  - [Status quo](#status-quo)
      - [Motivating Example](#motivating-example)
    - [Device plugins](#device-plugins)
      - [Possible POVs](#possible-povs)
  - [Kubeflow Changes](#kubeflow-changes)
    - [Change existing job related components that hard-code specific device limits ("nvidia.com/gpu")](#change-existing-job-related-components-that-hard-code-specific-device-limits-nvidiacomgpu)
    - [Update Profiles component](#update-profiles-component)
    - [Create a New Device Deployment component that can deploy specific devices via a DaemonSet](#create-a-new-device-deployment-component-that-can-deploy-specific-devices-via-a-daemonset)
    - [Create a New Device Monitoring component that can provide device specific metrics](#create-a-new-device-monitoring-component-that-can-provide-device-specific-metrics)
    - [Create a kube-batch component](#create-a-kube-batch-component)
    - [kfctl and platforms](#kfctl-and-platforms)
    - [Current components that are device specific](#current-components-that-are-device-specific)
    - [Relevant KEPS and Issues](#relevant-keps-and-issues)

### Scope

This is an umbrella issue for integrating device plugins into kubeflow. This touches on the following Kubeflow areas:
- kfctl (cluster creation, allocation of node pools)
- components where device information limits are specified
- profiles where extended resource quotas and limits are specified
- discovery/validation of resource availability prior to creating workflows (profiles)
- providing kustomize generators/transformers to add device info in kubeflow component containers for specific devices
- runtime device injection into containers (admission controllers), 
- application health/monitoring within the application (Application CRD) that can link to device monitoring url(s)

This issue will be a basis for follow-on CUJs, design docs, new issues within the areas noted above, PM involvement. Additional vetting and discussion will follow. The following sections may provide the basis for subsequent documents within the kubeflow public google drive.

### Status quo

Currently kubeflow does not have a story around different devices that may be required within existing components such as mxnet-job, openvino, pipeline, pytorch-job, tf-serving, tf-training...
Device specific hooks in container limits, volume mounts, daemonsets as described in [Device Plugins deployment](https://kubernetes.io/docs/concepts/extend-kubernetes/compute-storage-net/device-plugins/) are mostly specific to gpus. Successfully deploying a training or inference job that utilizes device hardware requires specific images to be used, with specific commands, such as [Intel Optimized for Tensorflow](https://software.intel.com/en-us/articles/intel-optimization-for-tensorflow-installation-guide?_ga=2.159497214.2133240433.1554058990-1030972731.1550631094#inpage-nav-5&elq_cid=1742718&erpm_id=4685615) within the pod container specification. Additionally for distributed training scenarios specific nodes may need to be gang scheduled by using PodGroups, PodDisruptionBudgets and specific taints and tolerations for the node and pod. Currently specifying these capabilities in kubeflow are limited.

##### Motivating Example

An example of a kubeflow issue opened last year was the following:
[TensorFlow Jupyter Notebook images 1.9 and above in gcr.io cannot see GPUs](https://github.com/kubeflow/kubeflow/issues/1828)

Here is what the user did in order to debug the problem

- checked if scheduling is occuring for GPU nodes

`kubectl get pods --output=wide`

- checked if limits on the container in the pod includes vendor-domain/device

`kubectl -n kubeflow describe pod jupyter-mike-2reeves-40xxxxxxxx-2enet`

```
Name: jupyter-mike-2reeves-40xxxxxxxx-2enet
Namespace: kubeflow
Node: gke-dsp-2-gpu-pool-165972e2-ks2n/10.142.0.4
Start Time: Fri, 19 Oct 2018 08:53:13 -0400
Labels: app=jupyterhub
component=singleuser-server
heritage=jupyterhub
Annotations: hub.jupyter.org/username=accounts.google.com:mike-2reeves-40xxxxxxxx-2enet
Status: Running
IP: 10.20.2.14
Containers:
notebook:
Container ID: docker://53ecaaf0dc32f21226934ac031b9273e31ede7c99d5a602a7fbdfb252428dc0f
Image: gcr.io/kubeflow-images-public/tensorflow-1.10.1-notebook-gpu:v0.3.0
Image ID: docker-pullable://gcr.io/kubeflow-images-public/tensorflow-1.10.1-notebook-gpu@sha256:24a32e6a533a6e7d46a649cbff247ebd86dee078d37489035d9bc440a9e2487f
Port: 8888/TCP
Host Port: 0/TCP
Args:
start-singleuser.sh
--ip="0.0.0.0"
--port=8888
--allow-root
State: Running
Started: Fri, 19 Oct 2018 08:53:23 -0400
Ready: True
Restart Count: 0
Limits:
nvidia.com/gpu: 1
Requests:
cpu: 1
memory: 2Gi
nvidia.com/gpu: 1
```
- Ran a device specific cli in the container process

`kubectl exec -it jupyter-mike-2reeves-40xxxxxxxx-2enet -- /bin/bash`

```
jovyan@jupyter-mike-2reeves-40xxxxxxxx-2enet:~$ nvidia-smi
Fri Oct 19 12:57:39 2018
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 396.26 Driver Version: 396.26 |
|-------------------------------+----------------------+----------------------+
| GPU Name Persistence-M| Bus-Id Disp.A | Volatile Uncorr. ECC |
| Fan Temp Perf Pwr:Usage/Cap| Memory-Usage | GPU-Util Compute M. |
|===============================+======================+======================|
| 0 Tesla K80 Off | 00000000:00:04.0 Off | 0 |
| N/A 30C P8 26W / 149W | 0MiB / 11441MiB | 0% Default |
+-------------------------------+----------------------+----------------------+

  +-----------------------------------------------------------------------------+
  | Processes:                                                       GPU Memory |
  |  GPU       PID   Type   Process name                             Usage      |
  |=============================================================================|
  |  No running processes found                                                 |
  +-----------------------------------------------------------------------------+`
```

- Ran python code in the container process

```
from tensorflow.python.client import device_lib

def get_available_devices():
   local_device_protos = device_lib.list_local_devices()
   return [x.name for x in local_device_protos]

print(get_available_devices())
```

#### Device plugins 

- [intel device plugins for kubernetes](https://github.com/intel/intel-device-plugins-for-kubernetes/blob/master/README.md)
- [Nvidia device plugin](https://github.com/NVIDIA/k8s-device-plugin)

##### Possible POVs

- [Intel GPU plugin](https://builders.intel.com/docs/networkbuilders/intel-device-plugins-for-kubernetes-appnote.pdf) Section 3
- Notebook that uses this GPU plugin 

#### Change existing job related components that hard-code specific device limits ("nvidia.com/gpu")

- chainer-job/prototypes/chainer-job.jsonnet
- katib/studyjobcontroller.libsonnet
- mxnet-job/prototypes/mxnet-job.jsonnet
- nvidia-inference-server/nvidia-inference-server.libsonnet
- profiles/sync-profile.jsonnet
- pytorch-job/prototypes/pytorch-job.jsonnet
- tf-batch-predict/tf-batch-predict.libsonnet
- tf-serving/tf-serving.libsonnet

Change above components to parameterize device type limits. This should be done using kustomize. 

#### Update Profiles component 

The Profiles CR provides a Resource Quota that is used when creating a Namespace. We need to make sure the Profile CRD Resource Quotas include extended resources like devices.

#### Create a New Device Deployment component that can deploy specific devices via a DaemonSet 

- [Device Plugin Deployment](https://kubernetes.io/docs/concepts/extend-kubernetes/compute-storage-net/device-plugins/#device-plugin-deployment)
- [Advertise Extended Resources for a Node](https://kubernetes.io/docs/tasks/administer-cluster/extended-resource-node/#advertise-a-new-extended-resource-on-one-of-your-nodes)
- [Nvidia device plugin daemonset](https://github.com/NVIDIA/k8s-device-plugin/blob/master/nvidia-device-plugin.yml)

#### Create a New Device Monitoring component that can provide device specific metrics 

 - [Kubelet Resource Metrics Endpoint
](https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/kubelet-resource-metrics-endpoint.md) 

#### Create a kube-batch component 

- See this [example](https://github.com/kubeflow/website/blob/master/content/docs/other-guides/job-scheduling.md) of a job requiring gang-scheduling

#### kfctl and platforms

Device integration begins with hardware availability which is specified during cluster creation.
For Kubeflow, the kfctl CLI specifies a platform flag eg: `kfctl init ~/kubeflow --platform gcp` which results in building a platform specific configuration under the specified directory ~/kubeflow.  The gcp platform includes cluster creation of gpu node pools. Both AWS and Azure platforms are being developed by contributors. Kubeflow has a yaml used by kfctl that could be extended for devices 

```
type NameValue struct {
	Name         string `json:"name,omitempty"`
	Value        string `json:"value,omitempty"`
	InitRequired bool   `json:"initRequired,omitempty"`
}
type Parameters map[string][]NameValue
// Default components configuration definitions.
type ComponentConfig struct {
	// Name of repository.
	Repo string `json:"repo,omitempty"`
	// List of default components.
	Components []string `json:"components,omitempty"`
	// List of default packages.
	Packages []string `json:"packages,omitempty"`
	// Parameters to be set to components using ks param set.
	ComponentParams Parameters `json:"componentParams,omitempty"`
	// Platform type.
	Platform string `json:"platform,omitempty"`
}
// KfDefSpec holds common attributes used by each platform
type KfDefSpec struct {
	config.ComponentConfig `json:",inline"`
	AppDir                 string `json:"appdir,omitempty"`
	Version                string `json:"version,omitempty"`
	MountLocal             bool   `json:"mountLocal,omitempty"`
	Project                string `json:"project,omitempty"`
	Email                  string `json:"email,omitempty"`
	IpName                 string `json:"ipName,omitempty"`
	Hostname               string `json:"hostname,omitempty"`
	Zone                   string `json:"zone,omitempty"`
	UseBasicAuth           bool   `json:"useBasicAuth"`
	SkipInitProject        bool   `json:"skipInitProject,omitempty"`
	UseIstio               bool   `json:"useIstio"`
	ServerVersion          string `json:"serverVersion,omitempty"`
	DeleteStorage          bool   `json:"deleteStorage,omitempty"`
}
```

Various 'flavors' of kfctl app.yamls are under [bootstrap/config](https://github.com/kubeflow/kubeflow/tree/master/bootstrap/config)

- kubeflow/deployment/gke/deployment_manager_configs/cluster.jinga
  Can this be updated so that the node-pool definitions are not gpu specific?

#### Current components that are device specific

- kubeflow/gcp/gpu-driver.libsonnet

#### Relevant KEPS and Issues

- [Kubelet Plugin Communication Establish Model](https://docs.google.com/document/d/1dtHpGY-gPe9sY7zzMGnm8Ywo09zJfNH-E1KEALFV39s/edit#heading=h.7fe6spexljh6)
- discussion of kube-arbitrator, extended resources see [kubeflow job scheduler](https://github.com/kubeflow/kubeflow/issues/68) 
- sig-node [Kubelet endpoint for device assignment observation details](https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/compute-device-assignment.md)
- sig-node [Kubelet Resource Metrics Endpoint](https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/kubelet-resource-metrics-endpoint.md)

**Comment by jlewi** [comment]
Thanks @kkasravi.

There's a lot of items here and I think CUJs will help us prioritize them. 

I would probably suggest starting with a notebook based CUJ.

/cc @vishh 

Related issues: 1457 GPU type selection in notebook manager

**Comment by kkasravi** [comment]
@jlewi I'll create a notebook based CUJ, thanks

**Comment by ashahba** [comment]
Adding @ashahba for review
